File: agents/universal-agent/internal/tracing/tracer.go
package tracing

import (
	"context"
	"fmt"
	"net/url"
	"os"

	"go.opentelemetry.io/otel/attribute"
	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracehttp"
	"go.opentelemetry.io/otel/sdk/resource"
	sdktrace "go.opentelemetry.io/otel/sdk/trace"
)

// InitTracer инициализирует OpenTelemetry TracerProvider с OTLP HTTP-экспортёром
// для отправки трейсов в Jaeger. Вызывающий обязан вызвать defer tp.Shutdown(ctx).
func InitTracer(serviceName string) (*sdktrace.TracerProvider, error) {
	ctx := context.Background()

	exporter, err := otlptracehttp.New(ctx,
		otlptracehttp.WithEndpoint(resolveEndpoint(os.Getenv("JAEGER_ENDPOINT"))),
		otlptracehttp.WithInsecure(),
	)
	if err != nil {
		return nil, fmt.Errorf("не удалось создать OTLP-экспортёр: %w", err)
	}

	res, err := resource.New(ctx,
		resource.WithAttributes(
			attribute.String("service.name", serviceName),
			attribute.String("service.version", "1.0.0"),
		),
	)
	if err != nil {
		_ = exporter.Shutdown(ctx)
		return nil, fmt.Errorf("не удалось создать resource: %w", err)
	}

	tp := sdktrace.NewTracerProvider(
		sdktrace.WithBatcher(exporter),
		sdktrace.WithResource(res),
	)
	return tp, nil
}

// resolveEndpoint извлекает host:port из URL-строки, либо возвращает дефолтный адрес.
func resolveEndpoint(raw string) string {
	if raw == "" {
		return "jaeger:4318"
	}
	u, err := url.Parse(raw)
	if err != nil || u.Host == "" {
		return raw
	}
	return u.Host
}

File: orchestrator/tracer.py
import os
from urllib.parse import urlparse

from opentelemetry import trace
from opentelemetry.exporter.otlp.proto.http.trace_exporter import OTLPSpanExporter
from opentelemetry.sdk.resources import Resource
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import BatchSpanProcessor


def init_tracer(service_name: str) -> TracerProvider:
    """
    Инициализирует глобальный OpenTelemetry TracerProvider с OTLP HTTP-экспортёром.
    Извлекает хост из JAEGER_ENDPOINT и использует OTLP-порт 4318.
    Вызывать один раз при старте сервиса.
    """
    raw = os.getenv("JAEGER_ENDPOINT", "http://jaeger:4318")
    parsed = urlparse(raw)
    otlp_url = f"http://{parsed.hostname}:4318/v1/traces"

    resource = Resource.create({
        "service.name": service_name,
        "service.version": "1.0.0",
    })
    exporter = OTLPSpanExporter(endpoint=otlp_url)
    provider = TracerProvider(resource=resource)
    provider.add_span_processor(BatchSpanProcessor(exporter))
    trace.set_tracer_provider(provider)
    return provider
